plugin: Connected hook

cdecker opened the pull request:
This is a simpler example than the `htlc_accepted` hook that I'm still working on, but it already allows us to do quite some interesting things. For example it'd allow us to implement a drain-plugin that allows connections to and from peers we are still negotiating a close with, but reject all connections that could result in a new channel.

It also serves as an example on how hooks are implemented.

cdecker commented:
Need to debug the CI failure tomorrow, hence the WIP tag :wink:

cdecker commented:
Should be ready to be reciewed, turns out it was just the switch from `@plugin.method('init')` to `@plugin.init()` decorators.

Ping @rustyrussell

rustyrussell commented on lightningd/peer_control.c:
```diff
@@ -411,39 +412,51 @@ void channel_errmsg(struct channel *channel,
+enum peer_connected_hook_result {
+	peer_connected_continue,
+	peer_connected_disconnect,
+};
```
enum constants should be CAPS...  I know it seems shouty but there it is

rustyrussell commented on wire/fromwire.c:
```diff
@@ -253,7 +253,7 @@ REGISTER_TYPE_TO_HEXSTR(channel_id);
  */
 void derive_channel_id(struct channel_id *channel_id,
-		       struct bitcoin_txid *txid, u16 txout)
+		       const struct bitcoin_txid *txid, u16 txout)
```
:+1:

rustyrussell commented on lightningd/peer_control.c:
```diff
@@ -617,6 +617,21 @@ static void
+	} else {
+		json_add_null(stream, "channel");
+	}
+	json_object_end(stream); /* .peer */
```
I think this hook may be far too helpful.  Just hand it the peer id, address and the global and local features from this connection message; the rest it can get via an RPC call if it wants?

rustyrussell commented on lightningd/json.c:
```diff
@@ -299,6 +299,11 @@ void json_add_bool(struct json_stream *result, const char *fieldname, bool value
+void json_add_null(struct json_stream *stream, const char *fieldname)
+{
+	json_add_member(stream, fieldname, "null");
+}
```
This commit should precede the prior one, but moreover: I think the usage here is wrong.  If a field doesn't exist it shouldn't exist, not be set to 'null'.  That's how we handle every other optional field.

rustyrussell commented on lightningd/peer_control.c:
```diff
@@ -638,7 +638,20 @@ static struct peer_connected_hook_response *
+	if (json_tok_streq(buffer, resulttok, "continue"))
+		resp->result = peer_connected_continue;
+	else
+		resp->result = peer_connected_disconnect;
```
Probably best to be explicit about the alternative to continue, eg "disconnect"?  Otherwise typos can have bad effects, and its hard to add other return types in the future (eg. an error message to send or something?).

cdecker commented:
Thanks @rustyrussell for the feedback, I rebased and amended as follows:

 - Enums are now extra shouty (https://github.com/ElementsProject/lightning/pull/2341/commits/a1f6656e86b705e9e8947b92a4ff3ebfe7d90bea and https://github.com/ElementsProject/lightning/pull/2341/commits/4a71ec075902142b987dee7e380a9a28a8c436d0) :wink: 
 - The `connected_hook` is now less helpful, but contains 100% more local features (https://github.com/ElementsProject/lightning/pull/2341/commits/a09dc4941f56c1bbba4eb528e7490da3f0741b49)
 - Adjusted the docs to mention where to get additional info if the hook isn't helpful enough (https://github.com/ElementsProject/lightning/pull/2341/commits/82263617b1ddd3f17be36077dd25d6099c9e4963)

I totally agree that the hook call was too heavy :+1:

cdecker commented:
Interpreting the approval above as an ACK:

ACK 968e9127e255f75875e54330205cd0b354ca9d72
